### PR TITLE
Add support for Teensy 3.5 & 3.6 boards

### PR DESCRIPTION
**Pull request by swilliams-dev**
- Updates the core directory to the latest
- Introduces a boards subdirectory similar to stmhal
- Adds modmachine
- Adds TEENSY_3.5 and 3.6
- Adds uart support

### CODE REVIEW

**Comment by swilliams-dev** [comment]
Greetings @dhylands ,

Not certain why I emailed first...

Thank you for all the micropython work and guides; with them I've been able to flash it to a 3.6. The project I'll be using it for is an arcade control panel, lit by neopixels. CircuitPython has most of the libraries already included as well as 3.1 support: https://github.com/adafruit/circuitpython/tree/master/ports/teensy which may be an easier merge.

Sounds like https://github.com/micropython/micropython/pull/2513 was deprioritized... would it be insane for a C newbie to try and fork off circuitpython and try to merge in https://github.com/dhylands/micropython/tree/teensy-3.6 ?

My hope in contacting you is to figure out if this merge is a realistic before I start down the path and perhaps get some advice/resources/reccomendations. You put so much work into getting this working, I'd like to continue the progress even if in a different direction.

Kind regards,
- Steven

**Comment by swilliams-dev** [comment]
Looks like there are a few merge conflicts (understatement)... this is going to be fun :)

File: teensy/Makefile [hunk 1]
```diff
@@ -1,4 +1,16 @@
...
+ifeq ($(wildcard boards/$(BOARD)/.),)
+$(error Invalid BOARD specified)
+endif
+
+# If the build directory is not given, make it reflect the board name.
+BUILD ?= build-$(BOARD)
+
 include ../py/mkenv.mk
+-include mpconfigport.mk
+include boards/$(BOARD)/mpconfigboard.mk
```
Comment (swilliams-dev) [inline teensy/Makefile]: Guessing this is necessary to specify different boards. https://github.com/micropython/micropython/wiki/Board-Teensy-3.1-3.5-3.6

File: teensy/Makefile [hunk 2]
```diff
@@ -18,26 +30,41 @@ $(error USE_ARDUINO_TOOLCHAIN requires that ARDUINO be set)
...
 
+# If you use teensy_loader_cli
+# from https://github.com/PaulStoffregen/teensy_loader_cli.git
+# then you no longer need to install arduino and teensyduino.
+
 ifeq ($(USE_ARDUINO_TOOLCHAIN),1)
 $(info Using ARDUINO toolchain)
-CROSS_COMPILE = $(ARDUINO)/hardware/tools/arm-none-eabi/bin/arm-none-eabi-
+CROSS_COMPILE = $(ARDUINO)/hardware/tools/arm/bin/arm-none-eabi-
+USE_TEENSY_LOADER_CLI = 0
```
Comment (swilliams-dev) [inline teensy/Makefile]: Reference https://github.com/PaulStoffregen/teensy_loader_cli

File: teensy/Makefile [hunk 3]
```diff
@@ -18,26 +30,41 @@ $(error USE_ARDUINO_TOOLCHAIN requires that ARDUINO be set)
...
+USE_TEENSY_LOADER_CLI = 1
 endif
 
-CFLAGS_TEENSY = -DF_CPU=96000000 -DUSB_SERIAL -D__MK20DX256__
-CFLAGS_CORTEX_M4 = -mthumb -mtune=cortex-m4 -mcpu=cortex-m4 -msoft-float -mfloat-abi=soft -fsingle-precision-constant -Wdouble-promotion $(CFLAGS_TEENSY)
+CFLAGS_TEENSY = -DUSB_SERIAL -DARDUINO=10611
+CFLAGS_CORTEX_M = -mthumb -mtune=cortex-m4 -mcpu=cortex-m4  -fsingle-precision-constant -Wdouble-promotion
+CFLAGS_MCU_mk20 = $(CFLAGS_CORTEX_M) -D__MK20DX256__ -msoft-float -mfloat-abi=soft
+CFLAGS_MCU_mk64 = $(CFLAGS_CORTEX_M) -D__MK64FX512__ -mfloat-abi=hard -mfpu=fpv4-sp-d16
+CFLAGS_MCU_mk66 = $(CFLAGS_CORTEX_M) -D__MK66FX1M0__ -mfloat-abi=hard -mfpu=fpv4-sp-d16
```
Comment (swilliams-dev) [inline teensy/Makefile]: MK66FX1M0 is the 3.6 (mine)

File: teensy/Makefile [hunk 4]
```diff
@@ -125,7 +153,7 @@ SRC_TEENSY = $(addprefix core/,\
 
 OBJ = $(PY_O) $(addprefix $(BUILD)/, $(SRC_C:.c=.o) $(STM_SRC_C:.c=.o) $(STM_SRC_S:.s=.o) $(SRC_TEENSY:.c=.o))
 OBJ += $(addprefix $(BUILD)/, $(LIB_SRC_C:.c=.o))
-OBJ += $(BUILD)/pins_gen.o
+OBJ += $(BUILD)/pins_$(BOARD).o
```
Comment (swilliams-dev) [inline teensy/Makefile]: looks conditional based on board built.

File: teensy/boards/TEENSY_3.6/mpconfigboard.h [hunk 5]
```diff
@@ -0,0 +1,75 @@
+
+// The following would be from a board specific file, if one existed
+
+#define MICROPY_HW_BOARD_NAME       "Teensy-3.6"
```
Comment (swilliams-dev) [inline teensy/boards/TEENSY_3.6/mpconfigboard.h]: Most relevant to what I can test.

File: teensy/boards/TEENSY_3.6/pins.csv [hunk 6]
```diff
@@ -0,0 +1,86 @@
+D0,PTB16
```
Comment (swilliams-dev) [inline teensy/boards/TEENSY_3.6/pins.csv]: Wish I had the source material that this csv came from. Guessing the datasheet but if not...

File: teensy/core/HardwareSerial.h [hunk 7]
```diff
@@ -166,22 +266,30 @@ class HardwareSerial : public Stream
...
+	operator bool()			{ return true; }
 };
 extern HardwareSerial Serial1;
+extern void serialEvent1(void);
 
 class HardwareSerial2 : public HardwareSerial
 {
 public:
-	virtual void begin(uint32_t baud) { serial2_begin(BAUD2DIV(baud)); }
+	virtual void begin(uint32_t baud) { serial2_begin(BAUD2DIV2(baud)); }
```
Comment (kirkpatt) [inline teensy/core/HardwareSerial.h]: What's the purpose of this change?

File: teensy/teensy_hal.h [hunk 8]
```diff
@@ -1,6 +1,11 @@
+#include <stdbool.h>
 #include <mk20dx128.h>
```
Comment (kirkpatt) [inline teensy/teensy_hal.h]: You change this include to be "kinetis.h" everywhere else. Is this intentional here?

File: teensy/modpyb.c [hunk 9]
```diff
@@ -56,93 +56,12 @@
 #include "dac.h"
 #include "usb.h"
 #include "portmodules.h"
+#include "modmachine.h"
 
 /// \module pyb - functions related to the pyboard
 ///
 /// The `pyb` module contains specific functions related to the pyboard.
 
-/// \function bootloader()
```
Comment (swilliams-dev) [inline teensy/modpyb.c]: The below removals may be problematic, flagging.

File: teensy/mpconfigport.h [hunk 10]
```diff
@@ -98,42 +119,12 @@ __attribute__(( always_inline )) static inline mp_uint_t disable_irq(void) {
 #define MICROPY_BEGIN_ATOMIC_SECTION()     disable_irq()
 #define MICROPY_END_ATOMIC_SECTION(state)  enable_irq(state)
 
+#include "mpconfigboard.h"
+
 // We need to provide a declaration/definition of alloca()
 #include <alloca.h>
 
-// The following would be from a board specific file, if one existed
```
Comment (swilliams-dev) [inline teensy/mpconfigport.h]: I think these are handled in the per board configs... but good to add here if things break.

File: teensy/teensy_hal.h (hunk 8 above)
Comment (swilliams-dev) [reply teensy/teensy_hal.h]: Thanks for flagging, kinetis info looks to be here:
https://github.com/micropython/micropython/pull/2513

**Comment by swilliams-dev** [comment]
Compilation had a few errors, as expected. Mainly in modmachine.c
modmachine.c:205:64: error: 'pyb_wfi_obj' undeclared here (not in a function)
modmachine.c:211:64: error: 'pyb_disable_irq_obj' undeclared here (not in a function)
modmachine.c:212:64: error: 'pyb_enable_irq_obj' undeclared here (not in a function)

Looks like I am missing an include.

**Comment by swilliams-dev** [comment]
Looks like this might be a macros issue:
https://github.com/micropython/micropython/commit/4ebdb1f2b217410cdc1cee0e0c0da8fceb7627f2

**Comment by swilliams-dev** [comment]
VirtualBox:~/attempt/circuitpython/teensy$ sudo ARDUINO=~/Downloads/arduino-1.8.5 BOARD=TEENSY_3.6 make
Use make V=1, make V=2 or set BUILD_VERBOSE similarly in your environment to increase build verbosity.
Using toolchain from PATH
Create build-TEENSY_3.6/genhdr/pins.h
Generating build-TEENSY_3.6/genhdr/mpversion.h
GEN build-TEENSY_3.6/genhdr/qstr.i.last
QSTR updated
GEN build-TEENSY_3.6/genhdr/qstrdefs.generated.h

In file included from main.c:20:0:
servo.h:5:1: warning: data definition has no type or storage class
MP_DECLARE_CONST_FUN_OBJ(pyb_servo_set_obj);
^~~~~~~~~~~~~~~~~~~~~~~~
servo.h:5:1: warning: type defaults to 'int' in declaration of 'MP_DECLARE_CONST_FUN_OBJ' [-Wimplicit-int]
servo.h:5:1: warning: parameter names (without types) in function declaration
servo.h:6:1: warning: data definition has no type or storage class
MP_DECLARE_CONST_FUN_OBJ(pyb_pwm_set_obj);
^~~~~~~~~~~~~~~~~~~~~~~~
servo.h:6:1: warning: type defaults to 'int' in declaration of 'MP_DECLARE_CONST_FUN_OBJ' [-Wimplicit-int]
servo.h:6:1: warning: parameter names (without types) in function declaration
In file included from main.c:23:0:
../stmhal/pin.h:89:1: warning: data definition has no type or storage class
MP_DECLARE_CONST_FUN_OBJ(pin_init_obj);
^~~~~~~~~~~~~~~~~~~~~~~~
../stmhal/pin.h:89:1: warning: type defaults to 'int' in declaration of 'MP_DECLARE_CONST_FUN_OBJ' [-Wimplicit-int]
../stmhal/pin.h:89:1: warning: parameter names (without types) in function declaration
In file included from main.c:24:0:
modmachine.h:36:1: warning: data definition has no type or storage class
MP_DECLARE_CONST_FUN_OBJ(machine_info_obj);
^~~~~~~~~~~~~~~~~~~~~~~~
modmachine.h:36:1: warning: type defaults to 'int' in declaration of 'MP_DECLARE_CONST_FUN_OBJ' [-Wimplicit-int]
modmachine.h:36:1: warning: parameter names (without types) in function declaration
modmachine.h:37:1: warning: data definition has no type or storage class
MP_DECLARE_CONST_FUN_OBJ(machine_unique_id_obj);
^~~~~~~~~~~~~~~~~~~~~~~~
modmachine.h:37:1: warning: type defaults to 'int' in declaration of 'MP_DECLARE_CONST_FUN_OBJ' [-Wimplicit-int]
modmachine.h:37:1: warning: parameter names (without types) in function declaration
modmachine.h:38:1: warning: data definition has no type or storage class
MP_DECLARE_CONST_FUN_OBJ(machine_reset_obj);
^~~~~~~~~~~~~~~~~~~~~~~~
modmachine.h:38:1: warning: type defaults to 'int' in declaration of 'MP_DECLARE_CONST_FUN_OBJ' [-Wimplicit-int]
modmachine.h:38:1: warning: parameter names (without types) in function declaration
modmachine.h:39:1: warning: data definition has no type or storage class
MP_DECLARE_CONST_FUN_OBJ(machine_bootloader_obj);
^~~~~~~~~~~~~~~~~~~~~~~~
modmachine.h:39:1: warning: type defaults to 'int' in declaration of 'MP_DECLARE_CONST_FUN_OBJ' [-Wimplicit-int]
modmachine.h:39:1: warning: parameter names (without types) in function declaration
modmachine.h:40:1: warning: data definition has no type or storage class
MP_DECLARE_CONST_FUN_OBJ(machine_freq_obj);
^~~~~~~~~~~~~~~~~~~~~~~~
modmachine.h:40:1: warning: type defaults to 'int' in declaration of 'MP_DECLARE_CONST_FUN_OBJ' [-Wimplicit-int]
modmachine.h:40:1: warning: parameter names (without types) in function declaration
modmachine.h:41:1: warning: data definition has no type or storage class
MP_DECLARE_CONST_FUN_OBJ(machine_sleep_obj);
^~~~~~~~~~~~~~~~~~~~~~~~
modmachine.h:41:1: warning: type defaults to 'int' in declaration of 'MP_DECLARE_CONST_FUN_OBJ' [-Wimplicit-int]
modmachine.h:41:1: warning: parameter names (without types) in function declaration
modmachine.h:42:1: warning: data definition has no type or storage class
MP_DECLARE_CONST_FUN_OBJ(machine_deepsleep_obj);
^~~~~~~~~~~~~~~~~~~~~~~~
modmachine.h:42:1: warning: type defaults to 'int' in declaration of 'MP_DECLARE_CONST_FUN_OBJ' [-Wimplicit-int]
modmachine.h:42:1: warning: parameter names (without types) in function declaration
In file included from led.c:9:0:
../stmhal/pin.h:89:1: warning: data definition has no type or storage class
MP_DECLARE_CONST_FUN_OBJ(pin_init_obj);
^~~~~~~~~~~~~~~~~~~~~~~~
../stmhal/pin.h:89:1: warning: type defaults to 'int' in declaration of 'MP_DECLARE_CONST_FUN_OBJ' [-Wimplicit-int]
../stmhal/pin.h:89:1: warning: parameter names (without types) in function declaration
In file included from modmachine.c:29:0:
modmachine.h:36:1: warning: data definition has no type or storage class
MP_DECLARE_CONST_FUN_OBJ(machine_info_obj);
^~~~~~~~~~~~~~~~~~~~~~~~
modmachine.h:36:1: warning: type defaults to 'int' in declaration of 'MP_DECLARE_CONST_FUN_OBJ' [-Wimplicit-int]
modmachine.h:36:1: warning: parameter names (without types) in function declaration
modmachine.h:37:1: warning: data definition has no type or storage class
MP_DECLARE_CONST_FUN_OBJ(machine_unique_id_obj);
^~~~~~~~~~~~~~~~~~~~~~~~
modmachine.h:37:1: warning: type defaults to 'int' in declaration of 'MP_DECLARE_CONST_FUN_OBJ' [-Wimplicit-int]
modmachine.h:37:1: warning: parameter names (without types) in function declaration
modmachine.h:38:1: warning: data definition has no type or storage class
MP_DECLARE_CONST_FUN_OBJ(machine_reset_obj);
^~~~~~~~~~~~~~~~~~~~~~~~
modmachine.h:38:1: warning: type defaults to 'int' in declaration of 'MP_DECLARE_CONST_FUN_OBJ' [-Wimplicit-int]
modmachine.h:38:1: warning: parameter names (without types) in function declaration
modmachine.h:39:1: warning: data definition has no type or storage class
MP_DECLARE_CONST_FUN_OBJ(machine_bootloader_obj);
^~~~~~~~~~~~~~~~~~~~~~~~
modmachine.h:39:1: warning: type defaults to 'int' in declaration of 'MP_DECLARE_CONST_FUN_OBJ' [-Wimplicit-int]
modmachine.h:39:1: warning: parameter names (without types) in function declaration
modmachine.h:40:1: warning: data definition has no type or storage class
MP_DECLARE_CONST_FUN_OBJ(machine_freq_obj);
^~~~~~~~~~~~~~~~~~~~~~~~
modmachine.h:40:1: warning: type defaults to 'int' in declaration of 'MP_DECLARE_CONST_FUN_OBJ' [-Wimplicit-int]
modmachine.h:40:1: warning: parameter names (without types) in function declaration
modmachine.h:41:1: warning: data definition has no type or storage class
MP_DECLARE_CONST_FUN_OBJ(machine_sleep_obj);
^~~~~~~~~~~~~~~~~~~~~~~~
modmachine.h:41:1: warning: type defaults to 'int' in declaration of 'MP_DECLARE_CONST_FUN_OBJ' [-Wimplicit-int]
modmachine.h:41:1: warning: parameter names (without types) in function declaration
modmachine.h:42:1: warning: data definition has no type or storage class
MP_DECLARE_CONST_FUN_OBJ(machine_deepsleep_obj);
^~~~~~~~~~~~~~~~~~~~~~~~
modmachine.h:42:1: warning: type defaults to 'int' in declaration of 'MP_DECLARE_CONST_FUN_OBJ' [-Wimplicit-int]
modmachine.h:42:1: warning: parameter names (without types) in function declaration
In file included from modmachine.c:36:0:
../stmhal/irq.h:73:1: warning: data definition has no type or storage class
MP_DECLARE_CONST_FUN_OBJ(pyb_wfi_obj);
^~~~~~~~~~~~~~~~~~~~~~~~
../stmhal/irq.h:73:1: warning: type defaults to 'int' in declaration of 'MP_DECLARE_CONST_FUN_OBJ' [-Wimplicit-int]
../stmhal/irq.h:73:1: warning: parameter names (without types) in function declaration
../stmhal/irq.h:74:1: warning: data definition has no type or storage class
MP_DECLARE_CONST_FUN_OBJ(pyb_disable_irq_obj);
^~~~~~~~~~~~~~~~~~~~~~~~
../stmhal/irq.h:74:1: warning: type defaults to 'int' in declaration of 'MP_DECLARE_CONST_FUN_OBJ' [-Wimplicit-int]
../stmhal/irq.h:74:1: warning: parameter names (without types) in function declaration
../stmhal/irq.h:75:1: warning: data definition has no type or storage class
MP_DECLARE_CONST_FUN_OBJ(pyb_enable_irq_obj);
^~~~~~~~~~~~~~~~~~~~~~~~
../stmhal/irq.h:75:1: warning: type defaults to 'int' in declaration of 'MP_DECLARE_CONST_FUN_OBJ' [-Wimplicit-int]
../stmhal/irq.h:75:1: warning: parameter names (without types) in function declaration
../stmhal/irq.h:76:1: warning: data definition has no type or storage class
MP_DECLARE_CONST_FUN_OBJ(pyb_irq_stats_obj);
^~~~~~~~~~~~~~~~~~~~~~~~
../stmhal/irq.h:76:1: warning: type defaults to 'int' in declaration of 'MP_DECLARE_CONST_FUN_OBJ' [-Wimplicit-int]
../stmhal/irq.h:76:1: warning: parameter names (without types) in function declaration
In file included from modmachine.c:37:0:
../stmhal/pin.h:89:1: warning: data definition has no type or storage class
MP_DECLARE_CONST_FUN_OBJ(pin_init_obj);
^~~~~~~~~~~~~~~~~~~~~~~~
../stmhal/pin.h:89:1: warning: type defaults to 'int' in declaration of 'MP_DECLARE_CONST_FUN_OBJ' [-Wimplicit-int]
../stmhal/pin.h:89:1: warning: parameter names (without types) in function declaration
modmachine.c:205:64: error: 'pyb_wfi_obj' undeclared here (not in a function); did you mean 'pyb_uart_obj_t'?
{ MP_OBJ_NEW_QSTR(MP_QSTR_idle),                (mp_obj_t)&pyb_wfi_obj },
^~~~~~~~~~~
pyb_uart_obj_t
modmachine.c:211:64: error: 'pyb_disable_irq_obj' undeclared here (not in a function); did you mean '__disable_irq'?
{ MP_OBJ_NEW_QSTR(MP_QSTR_disable_irq),         (mp_obj_t)&pyb_disable_irq_obj },
^~~~~~~~~~~~~~~~~~~
__disable_irq
modmachine.c:212:64: error: 'pyb_enable_irq_obj' undeclared here (not in a function); did you mean 'pyb_disable_irq_obj'?
{ MP_OBJ_NEW_QSTR(MP_QSTR_enable_irq),          (mp_obj_t)&pyb_enable_irq_obj },
^~~~~~~~~~~~~~~~~~
pyb_disable_irq_obj
../py/mkrules.mk:53: recipe for target 'build-TEENSY_3.6/modmachine.o' failed
make: *** [build-TEENSY_3.6/modmachine.o] Error 1

**Comment by swilliams-dev** [comment]
Compiles fine from https://github.com/dhylands/micropython/tree/teensy-3.6

File: teensy/modpyb.c [hunk 11]
```diff
@@ -56,93 +56,12 @@
 #include "dac.h"
 #include "usb.h"
 #include "portmodules.h"
+#include "modmachine.h"
```
Comment (swilliams-dev) [inline teensy/modpyb.c]: Think this includes the removed... but still getting *CONST_FUN_OBJ ref errors during compile, so maybe not? Or something to do with macro?
